### PR TITLE
feature blog: contextual logging update for 1.29

### PR DESCRIPTION
**Pull request by mengjiao-liu**
https://github.com/kubernetes/enhancements/issues/3077

see the [preview page](https://deploy-preview-43579--kubernetes-io-main-staging.netlify.app/blog/2023/12/06/contextual-logging-in-kubernetes-1-29/)

### CODE REVIEW

File: content/en/blog/_posts/2023-12-06-contextual-logging-in-kubernetes-1-29.md
Comment (pohly) [inline content/en/blog/_posts/2023-12-06-contextual-logging-in-kubernetes-1-29.md:14]:
```suggestion
two components (kube-scheduler and kube-controller-manager)
```

File: content/en/blog/_posts/2023-12-06-contextual-logging-in-kubernetes-1-29.md
Comment (pohly) [inline content/en/blog/_posts/2023-12-06-contextual-logging-in-kubernetes-1-29.md:63]:
```suggestion
it was important to be choose carefully when to add additional information:
```

File: content/en/blog/_posts/2023-12-06-contextual-logging-in-kubernetes-1-29.md
Comment (pohly) [inline content/en/blog/_posts/2023-12-06-contextual-logging-in-kubernetes-1-29.md:64]:
```suggestion
- At -v3 or lower, only `WithValues("pod")` is used once per scheduling cycle. 
```

File: content/en/blog/_posts/2023-12-06-contextual-logging-in-kubernetes-1-29.md
Comment (pohly) [inline content/en/blog/_posts/2023-12-06-contextual-logging-in-kubernetes-1-29.md:67]:
```suggestion
- At -v4 or higher, richer log entries get produced where `WithValues` is also used for the node (when applicable)
```

File: content/en/blog/_posts/2023-12-06-contextual-logging-in-kubernetes-1-29.md
Comment (pohly) [inline content/en/blog/_posts/2023-12-06-contextual-logging-in-kubernetes-1-29.md:68]:
```suggestion
  and `WithName` is used for the current operation and plugin.
```

**Review by pohly — CHANGES_REQUESTED** [review]
Looks good to me. I added some minor wording and markup suggestions.

**Comment by pohly** [comment]
/retitle feature blog: contextual logging update for 1.29

File: content/en/blog/_posts/2023-12-06-contextual-logging-in-kubernetes-1-29.md
Comment (mengjiao-liu) [reply content/en/blog/_posts/2023-12-06-contextual-logging-in-kubernetes-1-29.md]: Done.

**Comment by mengjiao-liu** [comment]
> Hi @mengjiao-liu how are you? here Communication Team 1.29, the deadline to the feature blog be ready to review is this Friday, Nov 17th.

Update the PR  status from draft to review.

**Comment by sftim** [comment]
/hold

I'm afraid this PR should target the _main_ branch. Please change it accordingly.

We'll also leave the hold in place until we know the publication date that release comms have assigned, and this PR matches that date.

**Comment by mengjiao-liu** [comment]
> I'm afraid this PR should target the main branch. Please change it accordingly.

However, the migration to kube-scheduler was done in 1.29, and the main branch is still 1.28.


> We'll also leave the hold in place until we know the publication date that release comms have assigned, and this PR matches that date.

I agree, this date is just a reference to [the 1.29 release date](https://www.kubernetes.dev/resources/release/).

**Comment by mengjiao-liu** [comment]
> We'll merge this PR to _main_, **after** v1.29 has released. You should change the target branch.

oops, I confused feature and blog PR. I will change the target branch.

File: content/en/blog/_posts/2023-12-06-contextual-logging-in-kubernetes-1-29.md
Comment (tengqm) [inline content/en/blog/_posts/2023-12-06-contextual-logging-in-kubernetes-1-29.md:92]:
```suggestion
In Kubernetes 1.29, enabling contextual logging at production verbosity (`-v3` or lower) 
```

File: content/en/blog/_posts/2023-12-06-contextual-logging-in-kubernetes-1-29.md
Comment (tengqm) [inline content/en/blog/_posts/2023-12-06-contextual-logging-in-kubernetes-1-29.md:93]:
```suggestion
caused no measurable slowdown for the `kube-scheduler` and is not expected for the
`kube-controller-manager` either. 
```

Comment (mengjiao-liu) [reply content/en/blog/_posts/2023-12-06-contextual-logging-in-kubernetes-1-29.md]: Done. Thanks for the catch!

File: content/en/blog/_posts/2023-12-06-contextual-logging-in-kubernetes-1-29.md
Comment (mengjiao-liu) [reply content/en/blog/_posts/2023-12-06-contextual-logging-in-kubernetes-1-29.md]: Done.

**Comment by krol3** [comment]
@kubernetes/sig-docs-blog-owners: Blog scheduled: Dec 20th, Publication Order Nro:9

**Review by sftim — COMMENTED** [review]
Thanks for proposing this article.

This PR has missed the deadline for PRs to be ready for review; however, part of that looks like a missed step from release comms.

You should promptly fix the publication date. The article is at risk of being omitted from post-release comms.

**Comment by mengjiao-liu** [comment]
> You should promptly fix the publication date. 

Sorry for missing some notifications due to something else, I have fixed the publication date.

**Comment by krol3** [comment]
> Blog scheduled: Dec 20th, Publication Order Nro:9

Hi @mengjiao-liu the comms 1.29 team will try to help with the review, any other help. Please let's know cc: @kubernetes/release-team-comms

File: content/en/blog/_posts/2023-12-20-contextual-logging-in-kubernetes-1-29.md
Comment (kcmartin) [inline content/en/blog/_posts/2023-12-20-contextual-logging-in-kubernetes-1-29.md:24]:
```suggestion
The binary decides the logging implementation, not the libraries. 
```

File: content/en/blog/_posts/2023-12-20-contextual-logging-in-kubernetes-1-29.md
Comment (kcmartin) [inline content/en/blog/_posts/2023-12-20-contextual-logging-in-kubernetes-1-29.md:34]:
```suggestion
  When passing this extended logger into a function, and the function uses it
```

File: content/en/blog/_posts/2023-12-20-contextual-logging-in-kubernetes-1-29.md
Comment (kcmartin) [inline content/en/blog/_posts/2023-12-20-contextual-logging-in-kubernetes-1-29.md:38]:
```suggestion
  all log entries for a certain operation, because the output from different operations gets interleaved.
```

File: content/en/blog/_posts/2023-12-20-contextual-logging-in-kubernetes-1-29.md
Comment (kcmartin) [inline content/en/blog/_posts/2023-12-20-contextual-logging-in-kubernetes-1-29.md:41]:
```suggestion
  Then, when a test fails, only the log output of the failed test gets shown by `go test`.
```

File: content/en/blog/_posts/2023-12-20-contextual-logging-in-kubernetes-1-29.md
Comment (kcmartin) [inline content/en/blog/_posts/2023-12-20-contextual-logging-in-kubernetes-1-29.md:47]:
```suggestion
it is *part* of the context, and not just *using* it. A practical advantage is that many APIs 
```

File: content/en/blog/_posts/2023-12-20-contextual-logging-in-kubernetes-1-29.md
Comment (sftim) [inline content/en/blog/_posts/2023-12-20-contextual-logging-in-kubernetes-1-29.md:8]:
```suggestion
**Authors**: [Mengjiao Liu](https://github.com/mengjiao-liu/) (DaoCloud), [Patrick Ohly](https://github.com/pohly) (Intel)
```

File: content/en/blog/_posts/2023-12-20-contextual-logging-in-kubernetes-1-29.md
Comment (sftim) [inline content/en/blog/_posts/2023-12-20-contextual-logging-in-kubernetes-1-29.md:18]:
```suggestion
## What is contextual logging?
```

File: content/en/blog/_posts/2023-12-20-contextual-logging-in-kubernetes-1-29.md
Comment (sftim) [inline content/en/blog/_posts/2023-12-20-contextual-logging-in-kubernetes-1-29.md:105]:
```suggestion
## Further reading
```

File: content/en/blog/_posts/2023-12-20-contextual-logging-in-kubernetes-1-29.md
Comment (sftim) [inline content/en/blog/_posts/2023-12-20-contextual-logging-in-kubernetes-1-29.md:110]:
```suggestion
## Get involved
```

**Review by sftim — COMMENTED** [review]
I'd prioritized reviews for the PRs that were in by the deadline, so I hadn't spotted this before:

- how about proposing this to https://kubernetes.dev/ blog?

If you're happy with that idea @mengjiao-liu, the Kubernetes blog team can make it happen - we can use our maintainer access to tweak the article before publication.
We can also mirror the article to https://k8s.io/blog/ even when the canonical URL is on the contributor site.

File: content/en/blog/_posts/2023-12-20-contextual-logging-in-kubernetes-1-29.md
Comment (kcmartin) [inline content/en/blog/_posts/2023-12-20-contextual-logging-in-kubernetes-1-29.md:48]:
```suggestion
already have a `ctx` parameter or can add one. This provides additional advantages, like being able to
```

File: content/en/blog/_posts/2023-12-20-contextual-logging-in-kubernetes-1-29.md
Comment (kcmartin) [inline content/en/blog/_posts/2023-12-20-contextual-logging-in-kubernetes-1-29.md:54]:
```suggestion
so it requires the `ContextualLogging` feature gate to be enabled.
```

File: content/en/blog/_posts/2023-12-20-contextual-logging-in-kubernetes-1-29.md
Comment (kcmartin) [inline content/en/blog/_posts/2023-12-20-contextual-logging-in-kubernetes-1-29.md:61]:
```suggestion
it is important to choose carefully when to add additional information:
```

File: content/en/blog/_posts/2023-12-20-contextual-logging-in-kubernetes-1-29.md
Comment (kcmartin) [inline content/en/blog/_posts/2023-12-20-contextual-logging-in-kubernetes-1-29.md:62]:
```suggestion
- At `-v3` or lower, only `WithValues("pod")` is used once per scheduling cycle.
```

File: content/en/blog/_posts/2023-12-20-contextual-logging-in-kubernetes-1-29.md
Comment (kcmartin) [inline content/en/blog/_posts/2023-12-20-contextual-logging-in-kubernetes-1-29.md:65]:
```suggestion
- At `-v4` or higher, richer log entries get produced where `WithValues` is also used for the node (when applicable)
```

File: content/en/blog/_posts/2023-12-20-contextual-logging-in-kubernetes-1-29.md
Comment (kcmartin) [inline content/en/blog/_posts/2023-12-20-contextual-logging-in-kubernetes-1-29.md:72]:
```suggestion
`pod` and `node` are already logged as parameters in individual log calls in `kube-scheduler` code.
```

File: content/en/blog/_posts/2023-12-20-contextual-logging-in-kubernetes-1-29.md
Comment (kcmartin) [inline content/en/blog/_posts/2023-12-20-contextual-logging-in-kubernetes-1-29.md:75]:
```suggestion
log calls can be simplified to avoid repeating those values.
```

File: content/en/blog/_posts/2023-12-20-contextual-logging-in-kubernetes-1-29.md
Comment (kcmartin) [inline content/en/blog/_posts/2023-12-20-contextual-logging-in-kubernetes-1-29.md:77]:
```suggestion
In `kube-controller-manager`, `WithName` is used to add the user-visible controller name to log output, 
```

File: content/en/blog/_posts/2023-12-20-contextual-logging-in-kubernetes-1-29.md
Comment (kcmartin) [inline content/en/blog/_posts/2023-12-20-contextual-logging-in-kubernetes-1-29.md:82]:
```suggestion
The `logger=”garbage-collector-controller”` was added by the `kube-controller-manager` core 
```

File: content/en/blog/_posts/2023-12-20-contextual-logging-in-kubernetes-1-29.md
Comment (kcmartin) [inline content/en/blog/_posts/2023-12-20-contextual-logging-in-kubernetes-1-29.md:83]:
```suggestion
when instantiating that controller and appears in all of its log entries - at least as long as the code
```

File: content/en/blog/_posts/2023-12-20-contextual-logging-in-kubernetes-1-29.md
Comment (kcmartin) [inline content/en/blog/_posts/2023-12-20-contextual-logging-in-kubernetes-1-29.md:99]:
```suggestion
whether it is because developers work on the code or because of the ongoing conversion
```

**Review by kcmartin — CHANGES_REQUESTED** [review]
Hi, Kristin from the 1.29 Comms Team here. I have made some suggested changes for formatting and clarity.

File: content/en/blog/_posts/2023-12-20-contextual-logging-in-kubernetes-1-29.md
Comment (sftim) [reply content/en/blog/_posts/2023-12-20-contextual-logging-in-kubernetes-1-29.md:54]: Extra credit: hyperlink feature gate to https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/

**Comment by mengjiao-liu** [comment]
Thank @sftim @kcmartin for your suggestions, I have addressed these comments.

**Comment by mengjiao-liu** [comment]
> how about proposing this to https://kubernetes.dev/ blog?
If you're happy with that idea @mengjiao-liu, the Kubernetes blog team can make it happen - we can use our maintainer access to tweak the article before publication.
We can also mirror the article to https://k8s.io/blog/ even when the canonical URL is on the contributor site.

WDYT? @pohly

**Comment by mengjiao-liu** [comment]
If I mirror this article, do I need to submit a PR in the https://github.com/kubernetes/contributor-site repo? Or is there anything else I need to do? @sftim

**Comment by sftim** [comment]
> If I mirror this article, do I need to submit a PR in the https://github.com/kubernetes/contributor-site repo? Or is there anything else I need to do? @sftim

You can either:
1. Submit a PR to https://github.com/kubernetes/contributor-site based on this PR, and change _this_ request into a PR that mirrors the article on the contributor site.
2. Grant permission for the blog editors to do the equivalent changes on your behalf. This is a better option if you're not confident you'd know how to get it right, because we don't yet a have a document to explain the task (“adapt an article proposed for the main blog into a contributor blog article that is also mirrored to the main Kubernetes blog”).
   In my experience, people doing that for the first time often get at least one detail wrong; I certainly don't blame them.

I recommend option 2, because I assume you do want that outcome and would prefer the more reliable approach. If you don't want this article to appear on the contributor blog at all, don't pick option 1 or option 2 (and we'll need to have a discussion about what works).

**Comment by mengjiao-liu** [comment]
Ok, so I choose option 2, and  I'll just wait for the blog editors to do it（IIUC）.

**Comment by Priyankasaggu11929** [comment]
@sftim, just wanted to confirm per the suggestion in https://github.com/kubernetes/website/pull/43579#pullrequestreview-1765695888 and following discussion, is this blog going to be part of 1.29 post release comms? Or, will be reviewed and published outside release comms?

**Comment by sftim** [comment]
I'll take care of this. It'll appear like post release comms _and_ it'll go on the contributor site (which will, if you look carefully, be the official version).

**Comment by sftim** [comment]
(still on my To Do list)

**Comment by sftim** [comment]
OK to remove the hold once https://github.com/kubernetes/contributor-site/pull/465 is merged.

**Comment by sftim** [comment]
/lgtm
/approve

See https://github.com/kubernetes/website/pull/43579#issuecomment-1864855897

**Pull request merged**
